Improvements to std::array<T, 0>.

rnicholl-google opened the pull request:
In the unstable ABI, std::array<T, 0> now uses only 1 byte with empty base optimization. In the stable ABI, it now can be used in constexpr context like std::array<int, 0> a; a.fill(1); return a;

Fixes llvm/llvm-project#74375

philnik777 commented:
Please keep the ABI flag separate. This isn't required to fix the bug and should have it's own discussion.

rnicholl-google commented:
> After looking at this PR, I would suggest that we go with #74667 which fixes the bug without any ABI change, and without requiring non-conforming API changes.
> 
> We can then either drop this or transform this patch into a change that makes `std::array<T, 0>` be 1 byte in all cases -- although the benefit of that in isolation is limited.

I actually don't think anything in the spec requires array<T const,0> to be non-copy assignable. Since the type itself is non const, you'd expect it would be copyable, unless the aggregate is un-copyable?. But the standard isn't very clear. I can separate this out into two separate commits. However, I am sending some emails to the std-proposals mailing list to see if we can get clarity around if std::array<T const, 0> should be copyable.

rnicholl-google committed:
Improvements to std::array<T, 0>.

It now can be used in constexpr context like std::arary<int, 0> a; a.fill(1); return a;

Also fixes begin() and end() returning nullptr when this is disallowed by the standard.

Requires llm/llvm-project#56814 fixed first.

rnicholl-google commented:
Closing this because separating it into a new one without the other changes.
